src/stats/stats.ts

```diff
@@ -0,0 +1,85 @@
+import _ from "lodash";
+
+import { FrameEntryType, PlayerIndexedType, Frames, FramesType } from "./common";
+
+export interface StatComputer<T> {
+    setPlayerIndices(indices: PlayerIndexedType[]): void;
+    processFrame(newFrame: FrameEntryType, allFrames: FramesType): void;
+    fetch(): T;
+}
+
+export interface StatOptions {
+    processOnTheFly: boolean;
+}
+
+const defaultOptions: StatOptions = {
+    processOnTheFly: false,
+}
+
+export class Stats {
+    private options: StatOptions;
+    private lastProcessedFrame: number | null = null;
+    private lastFrame: number;
+    private frames: FramesType = {};
+    private opponentIndices = new Array<PlayerIndexedType>();
+    private allComputers = new Array<StatComputer<unknown>>();
+
+    public constructor(options?: StatOptions) {
+        this.options = options || defaultOptions;
+    }
+
+    public setPlayerIndices(indices: PlayerIndexedType[]): void {
+        this.opponentIndices = indices;
+        this.allComputers.forEach(comp => comp.setPlayerIndices(indices));
+    }
+
+    public register(computer: StatComputer<unknown>): void {
+        this.allComputers.push(computer);
+    }
+
+    public registerAll(computers: StatComputer<unknown>[]): void {
+        this.allComputers = this.allComputers.concat(computers);
+    }
+
+    public process(): void {
+        // Finish processing if we're not up to date
+        if (this.lastFrame === this.lastProcessedFrame) {
+            return;
+        }
+        if (this.opponentIndices.length === 0) {
+            return;
+        }
+        let i = this.lastProcessedFrame ? this.lastProcessedFrame + 1 : Frames.FIRST;
+        while (Boolean(this.frames[i])) {
+            const frame = this.frames[i];
+            // Don't attempt to compute stats on frames that have not been fully received
+            if (!isCompletedFrame(this.opponentIndices, frame)) {
+                return;
+            }
+            this.allComputers.forEach(comp => comp.processFrame(frame, this.frames));
+            this.lastProcessedFrame = i;
+            i++;
+        }
+    }
+
+    public addFrame(frame: FrameEntryType): void {
+        this.frames[frame.frame] = frame;
```

Comment:
Feels weird to need this, we already have this in the location where `process` is called from SlippiGame. Why not just pass the frames in as an arg to `process` instead of re-managing the same object in this class?

Reply:
Not sure what this comment is referring to. But I definitely don't like the idea of passing in frames into process because that relies on the caller to have access to the frames. Where as with stats having its own state of the frames it knows where the processing is up to and it can just internally process the newer frames if needed.